cts/cts-regression.in

```diff
@@ -1,230 +1,257 @@
-            exit $CRM_EX_OK
-            ;;
-        -V|--verbose)
-            verbose="-V"
-            shift
-            ;;
-        -v|--valgrind)
-            valgrind="-v"
-            shift
-            ;;
-        scheduler|exec|fencing|cli)
-            add_test $1
-            shift
-            ;;
-        pacemaker_remote)
-            if [ $REMOTE_ENABLED -eq 0 ]; then
-                error "Pacemaker Remote not supported by this build"
-                exit $CRM_EX_USAGE
-            else
-                add_test $1
-                shift
-            fi
-            ;;
-        all)
-            add_test scheduler
-            add_test cli
-            add_test exec
-            add_test fencing
-            shift
-            ;;
-        *)
-            error "unknown option: $1"
-            echo
-            echo "$USAGE_TEXT"
-            exit $CRM_EX_USAGE
-            ;;
-    esac
-done
-
-if [ -z "$tests" ]; then
-    add_test scheduler
-    add_test cli
-fi
-
-run_tests $tests
+#!@PYTHON@
+"""Convenience wrapper for running Pacemaker regression tests.
+
+Usage: cts-regression [-h] [-V] [-v] [TEST ...]
+"""
+
+__copyright__ = 'Copyright 2012-2022 the Pacemaker project contributors'
+__license__ = 'GNU General Public License version 2 or later (GPLv2+) WITHOUT ANY WARRANTY'
+
+import argparse
+from enum import IntEnum, unique
+import os
+import stat
+import subprocess
+import sys
+import textwrap
+
+
+REMOTE_ENABLED = bool(os.getenv('@PC_NAME_GNUTLS@'))
+
+
+@unique
+class CrmExit(IntEnum):
+    """Pacemaker exit codes.
+
+    These values must be kept in sync with include/crm/common/results.h.
+    """
+
+    # Common convention
+    OK = 0
+    ERROR = 1
+
+    # LSB + OCF
+    INVALID_PARAM = 2
+    UNIMPLEMENT_FEATURE = 3
+    INSUFFICIENT_PRIV = 4
+    NOT_INSTALLED = 5
+    NOT_CONFIGURED = 6
+    NOT_RUNNING = 7
+    PROMOTED = 8
+    FAILED_PROMOTED = 9
+
+    # sysexits.h
+    USAGE = 64
+    DATAERR = 65
+    NOINPUT = 66
+    NOUSER = 67
+    NOHOST = 68
+    UNAVAILABLE = 69
+    SOFTWARE = 70
+    OSERR = 71
+    OSFILE = 72
+    CANTCREAT = 73
+    IOERR = 74
+    TEMPFAIL = 75
+    PROTOCOL = 76
+    NOPERM = 77
+    CONFIG = 78
+
+    # Custom
+    FATAL = 100
+    PANIC = 101
+    DISCONNECT = 102
+    OLD = 103
+    DIGEST = 104
+    NOSUCH = 105
+    QUORUM = 106
+    UNSAFE = 107
+    EXISTS = 108
+    MULTIPLE = 109
+    EXPIRED = 110
+    NOT_YET_IN_EFFECT = 111
+    INDETERMINATE = 112
+    UNSATISFIED = 113
+
+    # Other
+    TIMEOUT = 124
+
+    # OCF Resource Agent API 1.1
+    DEGRADED = 190
+    DEGRADED_PROMOTED = 191
+
+    # Custom
+    NONE = 193
+
+    MAX = 255
+
+
+class TestsArgAction(argparse.Action):
+    """Handle special cases for `tests` arguments.
+
+    Enforce a default value of ['cli', 'scheduler'].
+    Replace the 'all' alias with the tests that it represents.
+    Get rid of duplicates.
+
+    The main motivation is that when the `choices` argument of
```

Comment:
Just trust me on this haha... it was a pain and this is the cleanest thing I could come up with after consulting with the docs and StackOverflow. This type of thing is more or less the intended use case of subclassing `argparse.Action`.